### PR TITLE
Fix issue 2118

### PR DESCRIPTION
**Pull request by sklam**

### CODE REVIEW

File: numba/types/npytypes.py [hunk 1]
```diff
@@ -265,6 +265,10 @@ class Array(Buffer):
 
     def __init__(self, dtype, ndim, layout, readonly=False, name=None,
                  aligned=True):
+        if isinstance(dtype, DType):
```
Comment (pitrou) [inline numba/types/npytypes.py]: That doesn't really sound right. Perhaps we could do this in `__getitem__` where we handle the `dtype[:]` syntax, but I don't think we should let pass internal bugs where we give the wrong dtype when constructing an Array type instance.

**Comment by pitrou** [comment]
Note that bug #2118 is really a user error: `typeof(some numpy dtype)` returns the Numba type representing this dtype *object*. To get the Numba type of a record scalar, one has to pass a record scalar to `typeof()`.

Of course, we can still decide that `DType[:]` instantiates an array of the specific type denoted by the `DType`.

File: numba/tests/test_record_dtype.py [hunk 2]
```diff
@@ -861,6 +861,33 @@ def test_return_charseq_tuple(self):
             got = cfunc(self.nbsample1d, i)
             self.assertEqual(expected, got)
 
+class TestRecordDtypeTypeObj(unittest.TestCase):
+    def test_record_type_equiv(self):
```
Comment (pitrou) [inline numba/tests/test_record_dtype.py]: Such tests can go in `test_types.py`.

**Comment by sklam** [comment]
The `record_type[:]` usage in #2118 is from old examples and precede our introduction of `DType`.  Anyway, we should support it for usability.  I will adopt `DType[:]` style.

**Comment by sklam** [comment]
@pitrou, I've updated the patch

**Pull request merged**